ts/private/ts_project.bzl

```diff
@@ -184,81 +184,110 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
 
+    transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
+    transitive_inputs_depset = depset(
+        inputs,
+        transitive = transitive_inputs,
+    )
+
+    # tsc action for type-checking
     if use_tsc_noemit:
         # The type-checking action still need to produce some output, so we output the stdout
         # to a .typecheck file that ends up in the typecheck output group.
         typecheck_output = ctx.actions.declare_file(ctx.attr.name + ".typecheck")
         typecheck_outs.append(typecheck_output)
 
-        outputs.append(typecheck_output)
-        stdout_file = typecheck_output.path
-
+        typecheck_arguments = ctx.actions.args()
         if supports_workers:
-            arguments.add("--bazelValidationFile", typecheck_output.short_path)
+            typecheck_arguments.add("--bazelValidationFile", typecheck_output.short_path)
 
-        arguments.add("--noEmit")
+        typecheck_arguments.add("--noEmit")
+
+        ctx.actions.run(
+            executable = executable,
+            inputs = transitive_inputs_depset,
+            arguments = [arguments, typecheck_arguments],
+            outputs = [typecheck_output],
+            mnemonic = "TsProjectCheck",
+            execution_requirements = execution_requirements,
+            resource_set = resource_set(ctx.attr),
+            progress_message = "Type-checking TypeScript project %s [tsc -p %s]" % (
+                ctx.label,
+                tsconfig_path,
+            ),
+            env = {
+                "BAZEL_BINDIR": ctx.bin_dir.path,
+                "JS_BINARY__STDOUT_OUTPUT_FILE": typecheck_output.path,
+            },
+        )
     else:
         typecheck_outs.extend(output_types)
 
-    inputs_depset = depset()
-    if len(outputs) > 0:
-        transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
+    if use_tsc_for_js or use_tsc_for_dts:
+        tsc_emit_arguments = ctx.actions.args()
 
-        inputs_depset = depset(
-            copy_files_to_bin_actions(ctx, inputs),
-            transitive = transitive_inputs,
-        )
+        # Type-checking is done async as a separate action and can be skipped.
+        if ctx.attr.async_typecheck:
+            tsc_emit_arguments.add("--noCheck")
+            tsc_emit_arguments.add("--skipLibCheck")
+            tsc_emit_arguments.add("--noResolve")
+
+        if not use_tsc_for_js:
+            # Not emitting js
+            tsc_emit_arguments.add("--emitDeclarationOnly")
 
-        if ctx.attr.transpile != 0 and not ctx.attr.emit_declaration_only and not ctx.attr.no_emit:
-            if ctx.attr.declaration:
-                verb = "Transpiling & type-checking"
-            else:
-                verb = "Transpiling"
-        else:
-            verb = "Type-checking"
+        elif not use_tsc_for_dts:
+            # Not emitting declarations
+            # TODO: why doesn't this work with workers?
+            if not supports_workers:
```

Comment:
I'm in favor of dropping worker support in rules_ts 4.0. Not something we want to maintain and with async typecheck the need to keep large declaration graphs in memory for large typecheck actions is gone.

Reply:
FWIW isolatedDeclarations is no panacea. You can still end up with typecheck actions that take 30-90s especially if they have a big transitive tree.

I know you're not keen to maintain the worker but Jason and I were talking about the possibility of resurrecting it in a simpler form (check only, no emit). We (@MichaelMitchell-at) might do some work on this if we have cycles.

Another possibility is even if the worker is not doing --watch mode, we have found 20%+ performance improvement just from caching previously-parsed ASTs in memory instead of parsing from scratch for each project.

I guess my point is I'd be a fan of keeping the current level of (non)-support for workers a bit longer while we do a bit of investigation into how they could potentially help (or not) with isolatedDeclarations.